js/src/html/tokenizer.js

```diff
@@ -221,11 +226,43 @@ Tokenizer.prototype._read_open_handlebars = function(c, open_token) {
   return token;
 };
 
+Tokenizer.prototype._read_control_flows = function(c, open_token) {
+  var resulting_string = '';
+  var token = null;
+  // Only check for control flows if angular templating is set
+  if (!this._options.templating.includes('angular')) {
+    return token;
+  }
+
+  if (c === '@' && /[a-zA-Z0-9]/.test(this._input.peek(1))) {
```

Comment:
NOTE: This is on by default for HTML users.  This has the possibility of causing significant breakage for some users.  

We need to limit the chances of this triggering when not intentional. 

This should look for a known list of control flow words.  
https://angular.io/guide/control_flow seems to indicate the following list: `@if`, `@else`, `@for`,  `@empty`, `@switch`, `@case`, `@default`.  

Also, who are these control flow strings escaped? If I want to include a page that says `<p>Hello, to @if control flows!</p>`, how do I do that?

Reply:
Note that there are other keywords: https://angular.io/guide/defer

Reply:
@ssougnez 
Okay, I guess it's pretty obvious I'm obviously not an Angular user.

A more flexible pattern might be: 

```js
pattern_reader.starting_with(/\@[a-zA-z]+[ ]*/).until(/[({]/),
```

Still I'd rather lean toward a specific list of @-strings if possible.  

As an alternative, I'd also be okay with having this handler be off by default but that has it's own problems.

Reply:
To answer this question, we should have a look to see if it's possible to create custom control flow directives. It is possible to create custom structural directives (which are the "old" way of doing control flow) so it's not impossible that it is (will be) possible for control flow as well.

Even though I understand that a white list is more secure and easy to test/maintain, this would break if (when) angular allow this.

Reply:
> To answer this question, we should have a look to see if it's possible to create custom control flow directives. It is possible to create custom structural directives (which are the "old" way of doing control flow) so it's not impossible that it is (will be) possible for control flow as well.
> 
> Even though I understand that a white list is more secure and easy to test/maintain, this would break if (when) angular allow this.

That was my exact thought when I've written the code.




> NOTE: This is on by default for HTML users. This has the possibility of causing significant breakage for some users.
> 
> We need to limit the chances of this triggering when not intentional.
> 
> This should look for a known list of control flow words. https://angular.io/guide/control_flow seems to indicate the following list: `@if`, `@else`, `@for`, `@empty`, `@switch`, `@case`, `@default`.
> 
> You probably should us the pattern facilities already present in the library. This section shows who `CDATA` sections are checked for and read.
> 
> https://github.com/beautify-web/js-beautify/blob/1df08a0ca57b0c6a728223bbfd61e3ee81f447fe/js/src/html/tokenizer.js#L155-L165
> 
> The `cdata` pattern checks if the starting regex is present and then reads until it is finds the closing regex:
> 
> ```
> cdata: pattern_reader.starting_with(/<!\[CDATA\[/).until_after(/]]>/),
> ```
> 
> You don't need to read the whole control string using a Pattern but you can use to provide the opening guard. Something like:
> 
> ```
> angular_control_flow: pattern_reader.starting_with(/\@(if|else|for|empty|switch|case|default)[ ]*\(/).until(/\(/),
> ```
> 
> Keep the check for `c === '@'` to improve performance, but then call `read()` on that pattern. If it returns empty string, you have no match.

Your main point is that this beautifier is used in every HTML file, even if it's not an Angular file. That's why I added a new templating option. If `angular` is not set, then no control flow token will be set (so everything will remain as it was before this modification). If `angular` is set, then everything which is written after an `@` character, is treated as a control flow phrase.
![image](https://github.com/beautify-web/js-beautify/assets/84864554/99555ce1-819b-4053-bba9-436435bff3e3)
So it makes no sense to only match for the reserved keywords, as for Angular, everything is treated as control flow (even if it is a non-existing one).

But let me look into your other suggestions, I'll try to respond today.

Reply:
> ```js
> pattern_reader.starting_with(/\@[a-zA-z]+[ ]*/).until(/[({]/),
> ```

This solution marks only the `@` character and the control flow's name, but not the whole expression until it's opening curly brace.

Example:
<pre>
<b>@if</b>(condition) {
</pre>

The problem with this is that the <pre>@if<b>(condition) {</b></pre> part would be treated as simple text.
If the opening curly brace would be written in a new line, this would mess up the indentation.
```html
<!-- Unformatted -->
@if()
{
<div></div>
}

<!-- Formatted -->
@if()
    {
    <div></div>
}
```

Before commiting the solution I wrote, I tried to write a pattern, which selects the whole control flow opening part.
<pre><b>@if(condition) {</b></pre>
If we see just this simple example, we could say "read everything untilAfter an `{` character". That works, but not in all cases.
The problem with this is that, taking the `@if() {` for example, there can be an unknown amount of curly braces and parentheses inside the condition. So, if we had `@if({value: true} !== undefined) {`, then the pattern reader would only read until the first opening curly brace.
<pre><b>@if({</b>value: true} !== undefined) {</pre>
This is, unfortunately, definitely not correct. And to select the whole expression via REGEX does not seem possible (at least not with the ES version this library uses). That's why I implemented the selection in pure JS, counting the pairs of opening and closing parentheses, until they're the same amount and we get to the opening curly brace character.

Reply:
### Only a minor change, not a major one

> The problem with this is that, taking the @if() { for example, there can be an unknown amount of curly braces and parentheses inside the condition. So, if we had @if({value: true} !== undefined) {, then the pattern reader would only read until the first opening curly brace.

I'm sorry I should have explained this better. I'm not asking you to read the whole token using one call to `read()`.  I'm asking that you use the first call to `read()` to make your matching check stronger.  

I think the guard pattern we want is `pattern_reader.matching(/\@[a-zA-z]+[\n\t ]*[({]/)`.  This will only read if it finds this pattern, otherwise it returns blank and does not advance the input scanner. If it returns a matching string, then you can continue with the rest of your code with only minor changes. 

```suggestion
  // add to the __patterns initialization around line 72:
  // this.__patterns.angular_at_string_start = pattern_reader.matching(/\@[a-zA-z]+[\n\t ]*[({]/),
  if (c === '@') {
    resulting_string = this.__patterns.angular_at_string_start.read();
    if (resulting_string === '') {
      return token; // null, not a control flow at string
    }
    
    var opening_parentheses_count = resulting_string.endsWith('(') ? 1 : 0;
    var closing_parentheses_count = 0;

    // [ ... the rest of your implementation ...]
```


### On by default for HTML

> Your main point is that this beautifier is used in every HTML file, even if it's not an Angular file. That's why I added a new templating option. If `angular` is not set, then no control flow token will be set (so everything will remain as it was before this modification). 

It looks like this is what your code does, but it contradicts the behavior [described for the option](https://github.com/beautify-web/js-beautify/pull/2221/files#diff-19fdc95f956ffecf1fbe5bd670b218412313c00a31216a4ae0fe1b86193ed793R373):

```
--templating [...] auto = none in JavaScript, all in html
                                              ^^^^^^^^^^^
```

### What are our other options? 

* Add a new option called `indent-templating`.  It would be a selection list like `templating` but would default to `none` and have only two valid values `handlebars` and `angular`.  It would deprecate the current `indent_handlebars` setting - if that is set it adds  `handlebars` to `indent-templating`.  

* Do a major version bump and make some long needed changes including changing `templating` to default to `none`. Then we wouldn't have to worry about this causing unexpected effects.

Reply:
> I'm sorry I should have explained this better. I'm not asking you to read the whole token using one call to `read()`. I'm asking that you use the first call to `read()` to make your matching check stronger.
> 
> I think the guard pattern we want is `pattern_reader.matching(/\@[a-zA-z]+[\n\t ]*[({]/)`. This will only read if it finds this pattern, otherwise it returns blank and does not advance the input scanner. If it returns a matching string, then you can continue with the rest of your code with only minor changes.

I see, I didn't think of writing a pattern which doesn't match the whole token, but only a part of it. I'll do it asap. Thanks for the explanation!

> It looks like this is what your code does, but it contradicts the behavior [described for the option](https://github.com/beautify-web/js-beautify/pull/2221/files#diff-19fdc95f956ffecf1fbe5bd670b218412313c00a31216a4ae0fe1b86193ed793R373):
> 
> ```
> --templating [...] auto = none in JavaScript, all in html
>                                               ^^^^^^^^^^^
> ```
> 

Yes, I saw that this option will be on by default. I just haven't thought that this is the scope of this PR. But I understand that why it could cause conflicts and frustrate the users.

I can agree with both of your recommendations. The first sound like a quick fix, but it doesn't sound like a real solution, just a bandage to the current implementation. The latter is what I'd fancy, but it seems like that would require a bigger modification.

Now that I understand better your concerns, I might just suggest what you recommended in the first place, to add the reserved control flow keywords to the pattern reader (if, for, defer, etc), and only check for those, to minimize the conflicts. A follow-up task/issue should be to make those long needed changes to the `templating` option, then we could tokenize better. I'd be happy to do it in the future, but I cannot make any promises, that's why implementing the safer option now sounds a bit better.

What really bothers me in the current setup is that the `@`, `{`, `}` characters became reserved characters in angular (so you cannot use it in any other context, but only for control flow blocks), but they are normal characters in a non-angular scenario...